### PR TITLE
KAFKA-12718: SessionWindows are closed too early

### PR DESCRIPTION
**Pull request by gonzur**
Session windows should be closed only after `streamTime` reaches `windowEnd + gap + grace`.

### CODE REVIEW

**Comment by gonzur** [comment]
@mjsax pull request is up. Will be back later if there are any problems with the build.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java [hunk 1]
```diff
@@ -421,8 +421,8 @@ public void shouldLogAndMeterWhenSkippingLateRecordWithZeroGrace() {
         context.setRecordContext(new ProcessorRecordContext(0, -2, -3, "topic", new RecordHeaders()));
         processor.process("OnTime1", "1");
 
-        // dummy record to advance stream time = 1
-        context.setRecordContext(new ProcessorRecordContext(1, -2, -3, "topic", new RecordHeaders()));
+        // dummy record to advance stream time = 11
+        context.setRecordContext(new ProcessorRecordContext(10 + 1, -2, -3, "topic", new RecordHeaders()));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java]: Why `10 + 1`? Just make it `11`. Use the comment to explain the why

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java [hunk 2]
```diff
@@ -421,8 +421,8 @@ public void shouldLogAndMeterWhenSkippingLateRecordWithZeroGrace() {
         context.setRecordContext(new ProcessorRecordContext(0, -2, -3, "topic", new RecordHeaders()));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java:421]: Wondering if we should bump this timestamp to 10?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java [hunk 3]
```diff
@@ -489,16 +489,16 @@ public void shouldLogAndMeterWhenSkippingLateRecordWithNonzeroGrace() {
             context.setRecordContext(new ProcessorRecordContext(0, -2, -3, "topic", new RecordHeaders()));
             processor.process("OnTime1", "1");
 
-            // dummy record to advance stream time = 1
-            context.setRecordContext(new ProcessorRecordContext(1, -2, -3, "topic", new RecordHeaders()));
+            // dummy record to advance stream time = 11
+            context.setRecordContext(new ProcessorRecordContext(10 + 1, -2, -3, "topic", new RecordHeaders()));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java]: as above

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java [hunk 4]
```diff
@@ -489,16 +489,16 @@ public void shouldLogAndMeterWhenSkippingLateRecordWithNonzeroGrace() {
...
             processor.process("dummy", "dummy");
 
             // delayed record arrives on time, should not be skipped
             context.setRecordContext(new ProcessorRecordContext(0, -2, -3, "topic", new RecordHeaders()));
             processor.process("OnTime2", "1");
 
-            // dummy record to advance stream time = 2
-            context.setRecordContext(new ProcessorRecordContext(2, -2, -3, "topic", new RecordHeaders()));
+            // dummy record to advance stream time = 12
+            context.setRecordContext(new ProcessorRecordContext(10 + 2, -2, -3, "topic", new RecordHeaders()));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java]: as above

**Comment by gonzur** [comment]
@mjsax nits addressed and tests/build/arm working as of posting this. Leaving pinged for last round of review.

**Comment by mjsax** [comment]
Thanks for your PR! Merged to `trunk`.

**Comment by rackom** [comment]
Hello, is there any way how to fix it in version 2.5.0? Because this is causing us some issues with very short window + gap combinations (window: 10sec, grace: 5sec). It is not an option for us to upgrade to newer version since it has too many impacts on our solution atm.

**Comment by mjsax** [comment]
It's very unlikely that there will be another 2.5.x bug-fix release... Thus, upgrading to a newer version would be the only way to get the fix.

However, as a workaround, you can add the "gap" parameter to the grace-period manually to get the same effect. Thus, if you code is:
```
SessionWindows.gap(Duration.ofSeconds(5L)).grace(Duration.of(10L));
```
You can update to
```
SessionWindows.gap(Duration.ofSeconds(5L)).grace(Duration.of(15L));
```
as a workaround.

**Pull request merged**